### PR TITLE
EVG-6376 add permission checking middleware

### PR DESCRIPTION
**Pull request by john-m-liu**
This is a bit weird in that the test requires some mock constructs in packages that depend on the package with the actual middleware. So I had to either put the test in its own package, or the 2nd most logical package (I chose the latter)

### CODE REVIEW

File: auth_interfaces.go [hunk 1]
```diff
@@ -16,7 +16,7 @@ type User interface {
 	Username() string
 	GetAPIKey() string
 	Roles() []string
-	HasPermission(string, string, int) (bool, error)
+	HasPermission(string, string, string, int) (bool, error)
```
Comment (bsamek) [inline auth_interfaces.go]: This is now three string argument in a row. Can these go in a struct?

Comment (tychoish) [reply auth_interfaces.go]: +1

File: middleware_auth.go [hunk 2]
```diff
@@ -236,3 +236,76 @@ func (ra *restrictedAccessHandler) ServeHTTP(rw http.ResponseWriter, r *http.Req
...
 	rw.WriteHeader(http.StatusUnauthorized)
 }
+
+type requiresPermissionHandler struct {
+	opts RequiresPermissionMiddlewareOpts
+}
+
+type FindResourceFunc func(*http.Request) string
+
+type RequiresPermissionMiddlewareOpts struct {
```
Comment (tychoish) [inline middleware_auth.go]: docstrings?

File: middleware_auth.go [hunk 3]
```diff
@@ -236,3 +236,76 @@ func (ra *restrictedAccessHandler) ServeHTTP(rw http.ResponseWriter, r *http.Req
...
+	}
+
+	hasPermission, err := user.HasPermission(resource, rp.opts.ResourceType, rp.opts.PermissionKey, rp.opts.RequiredLevel)
+	if err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message":    "error checking permissions",
+			"user":       user.Username(),
+			"permission": rp.opts.PermissionKey,
+		}))
+	}
```
Comment (tychoish) [inline middleware_auth.go]: you don't need the ifstatment, wrapError noops sending if the err is nil.

File: middleware_auth.go [hunk 4]
```diff
@@ -236,3 +236,76 @@ func (ra *restrictedAccessHandler) ServeHTTP(rw http.ResponseWriter, r *http.Req
...
+	if hasPermission {
+		next(rw, r)
+		return
+	}
```
Comment (tychoish) [inline middleware_auth.go]: I would put the exceptional case (error, not authorized.) first with the early return followed by the next() call.

File: auth_interfaces.go (hunk 1 above)
Comment (john-m-liu) [reply auth_interfaces.go]: done

File: middleware_auth.go (hunk 2 above)
Comment (john-m-liu) [reply middleware_auth.go]: added

File: middleware_auth.go (hunk 3 above)
Comment (john-m-liu) [reply middleware_auth.go]: changed

File: middleware_auth.go (hunk 4 above)
Comment (john-m-liu) [reply middleware_auth.go]: changed

**Pull request merged**